### PR TITLE
Add XSOAR Contribution Design Tutorial

### PR DESCRIPTION
**Pull request by fvigo**
## Status
Ready

## Description
Added a new tutorial on design

### CODE REVIEW

File: docs/tutorials/tut-design.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+
+2. Incident enrichment: following the creation of an incident, there is usually an enrichment phase where additional data is automatically collected from multiple sources (reputation, directories, DBs, etc.) to provide the analyst with all the necessary context to make decisions on priority and impact.
+
+3. Incident response: once all the data is collected, it's time to handle the incident response. This could be as simple as closing it (in case it has been determined that is a false positive), to very rich and complex playbooks that include several automated (or semi-automated) remediation steps across the entire IT infrastructure, such as blocking, quarantining, notifying people, collecting even more data, forensics analysis, reporting, etc. The number of actions available out of the box through Cortex XSOAR via Integrations is huge, and more can be added through contributions. Everything that can be automated is a huge help for the SOC team.
+
+With this in mind, you could start thinking about how your contribution cool look like:
```
Comment (bakatzir) [inline docs/tutorials/tut-design.md:36]:
```suggestion
With this in mind, you could start thinking about how your contribution should look like:

```

File: docs/tutorials/tut-design.md [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+- Playbook Design
+- Content Pack Design
+
+Once all the above steps are completed, the coding part will be trivial. Let's get started.
+
+## Use Cases
+
+We are grateful that you decided to spend time and effort to contribute to Cortex XSOAR. At first, it's important to understand what you want to achieve with your contribution. We recommend to start thinking about User Stories, to determine what would constitute a successful outcome of your effort.
+
+Cortex XSOAR is a solution that can SOCs use to reduce manual activity by providing automation via playbooks and integrations with multiple third party APIs. The flexibility of the product is such as there isn't a specific script but, in very general terms, often we see the following phases of an incident lifecycle:
```
Comment (bakatzir) [inline docs/tutorials/tut-design.md:28]:
```suggestion
Cortex XSOAR is a solution that SOCs use to reduce manual activity by providing automation via playbooks and integrations with multiple third party APIs. The flexibility of the product is such as there isn't a specific script but, in very general terms, often we see the following phases of an incident lifecycle:

```

File: docs/tutorials/tut-design.md [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+Depending on whether you are a Palo Alto Networks Technology Partner, an Individual Contributor, a Customer or an Enthusiast and depending on how you want your contribution to appear on the [Cortex XSOAR Marketplace](../partners/marketplace), the thoroughness and complexity of your final product will be different.
+
+This section summarizes some guidelines to get started.
+
+### Playbooks
+
+If you have an end-to-end use case in mind, most likely you are going to build one or more **Playbooks**. Start thinking about the process that you want to automate and enumerate the steps and the decisions that are part of it: those will become your playbook's tasks. Figure out if all these building blocks are already part of Cortex XSOAR by browsing the Marketplace.
+
+You can have multiple Playbooks in the same Content Pack, as long as they are related to a similar end-to-end use case. If they are completely separate, consider splitting them in multiple Content Packs.
+
```
Comment (bakatzir) [inline docs/tutorials/tut-design.md:71]: add a few playbooks examples from the reference?
also, let's mention that we can use tasks from other packs as mandatory/non mandatory.
e.g: I'll do endpoint management: use XDR, CB, crowdStrike and cyclance packs in my pack.

File: docs/tutorials/tut-design.md [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+### Integrations
+
+You will need to create an **Integration** if the answer is yes to any of the following questions:
+- You have a use case in mind that requires communicating with a third party system or API, but there is no Integration available for it in the Cortex XSOAR Marketplace.
+- You are a vendor and you want Cortex XSOAR to be able to interact with your product or retrieve indicators from your Threat Intel Feed. 
+
+Think of Integrations as the building blocks that enable external communications and form the foundations of Playbooks.
```
Comment (bakatzir) [inline docs/tutorials/tut-design.md:78]: add that each integration should include api endpoint of only one product. so instance configuration will be easy. mention also the VENDOR-OBJECT-ACTION naming convention.

File: docs/tutorials/tut-design.md [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+SOCs analysts might be interested only in a subset of the incoming Alerts, so when they configure your integration in Cortex XSOAR, they expect to find Parameters that allow them to filter and determine what alerts are going to generate incidents or discarded.
+
+Common filters for fetching incidents are:
+ - **Maximum number of incidents per fetch**:
+ - **Type**
+ - **Severity**
+ - **First Fetch**
+
+
```
Comment (bakatzir) [inline docs/tutorials/tut-design.md:180]: #### Fetch Indicators Parameters
https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/commands/common/constants.py#L870

File: docs/tutorials/tut-design.md [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+These parameters determine how Cortex XSOAR connects to the third party API. 
+
+Usually there are at least two parameters:
+ - **url**: the URL Cortex XSOAR should connect to. In case the API is SaaS and the URL is the same for all customers and is not going to change over time, you can omit this parameter and hardcode the URL in the code. Some SaaS products need just the name of the tenant instead of the full URL (as they add the common suffix in the code). For example: `customer1.my-test-saas-service.info` and `customer2.my-test-sas-service.info`: in this case the input parameter could just be the name of the tenant (`customer1` or `customer`).
+ - **API Key**: usually credentials are needed to authenticate to the API you want to interact with. Many times this is in the form of an API Key, a secure quantity that gets stored in Cortex XSOAR and used in somewhere in your code (i.e. to populate HTTP headers) to craft the requests. Sometimes APIs use different authentication parameters: tokens, client_id and client_secret combination, and so on. Make sure you capture all the required parameters that will allow a machine-to-machine authentication between Cortex XSOAR and your API.
```
Comment (bakatzir) [inline docs/tutorials/tut-design.md:156]: or credentials

**Comment by anara123** [comment]
@michalgold @Arsenikr @ShahafBenYakir fyi

File: docs/tutorials/tut-design.md [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+ - Does the product provide any reputation on Indicators of Compromise (IOCs)?
+ - Does the product provide feeds of IOCs? 
+
+### Integration parameters
+
+Each integration has *parameters*, that are configured by the customers when they configure an instance on their Cortex XSOAR system.
+
+The following parameter types are supported: Short Text, Long Text, Boolean, Encrypted, Single Select, Multi Select.
+
+Even if XSOAR doesn't make any distinction, in this tutorial we group  parameters in three categories.
```
Comment (anara123) [inline docs/tutorials/tut-design.md:148]:
```suggestion
Even if XSOAR doesn't make any distinction, in this tutorial we group parameters in three categories.

```

File: docs/tutorials/tut-design.md [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+ - **API Key**: usually credentials are needed to authenticate to the API you want to interact with. Many times this is in the form of an API Key, a secure quantity that gets stored in Cortex XSOAR and used in somewhere in your code (i.e. to populate HTTP headers) to craft the requests. Sometimes APIs use different authentication parameters: tokens, client_id and client_secret combination, and so on. Make sure you capture all the required parameters that will allow a machine-to-machine authentication between Cortex XSOAR and your API.
+ 
+    **Note**: if you are using authentication mechanisms based on short-lived access tokens such as JSON Web Tokens (JWT) and long-lived refresh tokens, take a look at the [Integration Cache](../integrations/integration-cache) article.
+proxy
+insecure
```
Comment (anara123) [inline docs/tutorials/tut-design.md:160]: delete

File: docs/tutorials/tut-design.md [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+#### Fetch Incidents Parameters
+
+If your integration [fetches incidents](../integrations/fetching-incidents), you should define what type of entities in the third party product you are connecting to you are retrieving. Every product has its own nomenclature: they can be called Alerts, Incidents, etc: for the rest of this section we are going to assume they are called Alerts in the third party product, and they are mapped 1:1 to Cortex XSOAR incidents.
+
+Many products are very verbose, and can potentially generate lots of alerts of different types, with different level of severity. Each alert could have a status (i.e. `open` or `resolved`) as well as attributes.
+
+SOCs analysts might be interested only in a subset of the incoming Alerts, so when they configure your integration in Cortex XSOAR, they expect to find Parameters that allow them to filter and determine what alerts are going to generate incidents or discarded.
+
+Common filters for fetching incidents are:
+ - **Maximum number of incidents per fetch**:
```
Comment (anara123) [inline docs/tutorials/tut-design.md:175]:
```suggestion
 - **Maximum number of incidents per fetch**

```

File: docs/tutorials/tut-design.md [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+ 
+
+### Other Parameters
+First of all you must determine what your integration does
+|Functionality||
+|
+single per product
+atomic api calls
+[Code Conventions](../integrations/code-conventions)
+[PowerShell](../integrations/powershell-code)
```
Comment (anara123) [inline docs/tutorials/tut-design.md:189]: invalid formatting

File: docs/tutorials/tut-design.md [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+
+### Docker Image
+
+[Docker](../integrations/docker)
```
Comment (anara123) [inline docs/tutorials/tut-design.md:193]: do we want the docker image as part of the design?

**Comment by ShahafBenYakir** [comment]
@anara123 Any idea why we don't have netlify here?

**Comment by fvigo** [comment]
> Any idea why we don't have netlify here?

@ShahafBenYakir Should be building it now

File: docs/tutorials/tut-design.md (hunk 4 above)
Comment (fvigo) [reply docs/tutorials/tut-design.md:77]: done, also mentioned the conventions later in the doc

File: docs/tutorials/tut-design.md [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+
+Think of Integrations as the building blocks that enable external communications and form the foundations of Playbooks.
+
+While it's possible to have multiple integrations in a single Content Pack, we generally recommend to include only a single integration in each Content Pack (i.e. if you are building two integrations to interact with two different third party products, we recommend to create one Content Pack for each product)
+
+### Classifiers, Mappers, Types, Fields and Layouts
+
+If you are planning to create a new Integration that [fetches incidents](../integrations/fetching-incidents) from a third party system, it is recommended that you create a few additional items and add them to your Content Pack:
+- **Incident Types**: Cortex XSOAR ships with a number of out of the box Incident Types, however it is recommended to create one or more Incident Types that are specific to the product you are creating incidents from. More [info](../incidents/incident-types).
+- **Incident Fields**: Product-specific fields are important to create a data model that is relevant to the incident type you have created. More [info](../incidents/incident-fields).
```
Comment (bakatzir) [inline docs/tutorials/tut-design.md:85]:
```suggestion
- **Incident Fields**: Product-specific fields are important to create a data model that is relevant to the incident type you have created. All of the incident fields should be associated only to the incident type you have created, and their name should be prefixed accordingly to indicate it. More [info](../incidents/incident-fields).

```

**Review by bakatzir — APPROVED** [review]
1 last comment. great document.

**Comment by fvigo** [comment]
PR is now ready for review

File: docs/tutorials/tut-design.md [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+
+#### Command Outputs
+
+Every automation script and integration command returns several types of outputs:
+ - Human Readable: this is shown to the user in the War Room and is typically formatted in a way that is understandable by the SOC analyst. The Human Readable data is usually a subset of the entire information returned by your command.
```
Comment (anara123) [inline docs/tutorials/tut-design.md:253]: For Human Readable I would also add that it is a good practice to show the details the user see in the original product. So if in product Incident has 5 fields, then in Human Readable make sure to show these fields first

File: docs/tutorials/tut-design.md [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+
+- Use [Long Running](../integrations/long-running) containers: this is suitable for services that need to keep a connection open for long time or need to open a listening TCP port. For example [Slack](https://github.com/demisto/content/tree/master/Packs/Slack/Integrations/Slack).
+
+There are cases where you want to build commands that perform generic well-known actions that are common across several use cases, for example Reputation commands (i.e. commands that return enrichment and reputation information about indicators, such as IPs). For those scenarios, we have standardized ways to define command names, inputs and outputs that will make interoperability much easier. For more information, check out the docs about [Generic Commands](../integrations/generic-commands-about) and [DBotScore](../integrations/dbot).
+
+#### Command Arguments
+
+Arguments are the inputs of your integration commands. They can be mandatory or not, and can have default and predefined values. More details [here](../integrations/yaml-file#commands).
```
Comment (anara123) [inline docs/tutorials/tut-design.md:240]: For the arguments we should suggest a good practice of making the inputs similar to what user required to provide in the original product. So for example in the product to create an incident user should choose a severity from 3 options High, Medium and Low, but the api supports 1,2,3 (numbers). Then the integration should support High, Medium and Low and convert these values in the integration. The user experience of integration should be familiar to the user

**Review by anara123 — APPROVED** [review]
Looks great! See my comments

File: docs/tutorials/tut-design.md [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+
+Integration parameters should be named using [snake_case](https://medium.com/better-programming/string-case-styles-camel-pascal-snake-and-kebab-case-981407998841).
+
+#### Fetch Incidents Parameters
+
+If your integration [fetches incidents](../integrations/fetching-incidents), you should define what type of entities in the third party product you are connecting to you are retrieving. Every product has its own nomenclature: they can be called Alerts, Incidents, etc: for the rest of this section we are going to assume they are called Alerts in the third party product, and they are mapped 1:1 to Cortex XSOAR incidents.
+
+Many products are very verbose, and can potentially generate lots of alerts of different types, with different level of severity. Each alert could have a status (i.e. `open` or `resolved`) as well as attributes.
+
+SOCs analysts might be interested only in a subset of the incoming Alerts, so when they configure your integration in Cortex XSOAR, they expect to find Parameters that allow them to filter and determine what alerts are going to generate incidents or discarded.
```
Comment (Arsenikr) [inline docs/tutorials/tut-design.md:182]: i would also add a note about a free query string to filter incidents if the product allows it. 
users are used to the syntax and it allows freedom to search exact incidens

**Review by Arsenikr — CHANGES_REQUESTED** [review]

**Review by Arsenikr — COMMENTED** [review]
added some notes

**Review by Arsenikr — APPROVED** [review]

**Pull request merged**